runtime/lua/vim/keymap.lua

```diff
@@ -0,0 +1,127 @@
+local keymap = {}
+
+--- Add a new keymap.
+--- Examples:
+--- <pre>
+---   -- You can make keymaps worth lua functions
+---   vim.keymap.set('n', 'lhs', function() print("real lua function") end)
+---
+---   -- Can use it to map multiple modes
+---   vim.keymap.set({'n', 'v'}, '<leader>lr', vim.lsp.buf.references, { silent = true, buffer=true })
+---
+---   -- You can map a key for specific buffer
+---   vim.keymap.set('n', '<leader>w', "<cmd>w<cr>", { silent = true, buffer = 5 })
+---
+---   -- Also expr keymaps too. Notice the replace_termcodes when making lua expr maps.
+---   vim.keymap.set('i', '<Tab>', function()
+---     return vim.fn.pumvisible() == 1 and "<C-n>" or "<Tab>"
+---   end, { silent = true, expr = true, replace_termcodes = true })
+---   -- While setting <Plug> mapping remember to set remap to true
+---   vim.keymap.set('n', '[%', '<Plug>(MatchitNormalMultiBackward)', {silent = true, remap = true })
+--- </pre>
```

Comment:
I think it's worth mentioning an example of using `require` from a plugin. As @clason mentioned this will be a common use case and it's likely many people will just attempt to use `require("foo").some_func` which of course will load the `foo` plugin when the keymap is defined (which isn't what people want). Having a canonical example of using `function() require("foo").some_func() end` would be good.

Reply:
Agreed; much of the issue is that we simply haven't worked out best practices around these new Lua variants of well-established vimscript patterns. (@nanotee to the rescue ;))

I'll have to play around whether `loadstring` or anonymous functions are the more convenient pattern (or what the exact trade-offs are for each); in any case, both are supported in the current API, so any further improvement can -- and should! -- come later.

(The _real_ drawback with the `function()` thing is that `stylua` insists on making it multiline...)

Reply:
you could also do `require_on_call` (https://github.com/tjdevries/lazy.nvim/blob/master/lua/lazy.lua#L49-L72) but people will have different preferences on how to do so I think.

Reply:
I think "lazy require" is a good idea but independent from this PR (as in, I don't see how the implementation or API would change if that were available), so it's better to discuss this as part of the general "require goes *brrrr*" discussion around caching etc., for example in https://github.com/neovim/neovim/pull/15436.

Mentioning that Lua functions are actually accessed during `set` (so `require` is executed) and mentioning the `function` wrapper workaround is enough here. I'd prefer it if this PR got merged early and then iterated on during the 0.7 development cycle.